### ISSUE TITLE
ACP: add `std::any::is_same_type::<T, U>() -> bool`

### ISSUE THREAD

**Issue by tgross35** [rust-lang/libs-team#411]
# Proposal

## Problem statement

https://github.com/rust-lang/libs-team/issues/231 was recently accepted as a way to check if two types are the same in `const` contexts, providing a new `TypeId::matches`. However, this API suffers from a few problems:

1. Its goal is to be useful at compile time, but it relies on being able to construct `TypeId`s. Whether or not we can provide a safe `const TypeId::of` is not well known at this time; if it is possible, it will likely be a while off. (https://github.com/rust-lang/rust/issues/77125, https://github.com/rust-lang/rust/issues/10389)
2. If we eventually get a const `PartialEq`, `TypeId::matches` will become a duplicate of a more natural API.

These two problems seem to provide significant barriers to eventual stabiliztion of `TypeId::matches`

## Motivating examples or use cases

Most of the use cases for const type equality checks is comparing types that are available either as a generic or as a concrete type, rather than unknown `TypeId`s. This is currently true by default since `const TypeId::of` is not available, but seems likely to be the case more generally.

One use case of checking types for equality is to provide downcasting behavior:

```rust
use core::ptr;

const fn const_downcast<T, Expected>(value: &T) -> Option<&Expected> {
    if /* magic expression */ {
        Some( unsafe { &*ptr::from_ref(value).cast::<Expected>() })
    } else {
        None
    }
}
```

https://github.com/rust-lang/libs-team/issues/231 provides a more complete example.

## Solution sketch

Rather than provide an interface that relies on `TypeId`s, provide a standalone function that takes the types as generic parameters:

```rust
// core::any

/// Return `true` if `T` and `U` are the same type.
const fn is_same_type<T: 'static, U: 'static>() -> bool {
    // or intrinsic call
    TypeId::of::<T>().t == TypeId::of::<U>().t
}
```

Usage:

```rust
use core::{any, ptr};

const fn const_downcast<T, Expected>(value: &T) -> Option<&Expected> {
    if any::is_same_type::<T, Expected>() {
        Some( unsafe { &*ptr::from_ref(value).cast::<Expected>() })
    } else {
        None
    }
}
```

This can be implemented using `TypeId` (as in this example), but it could also be implemented as an intrinsic. The intrinsic option allows hooking into compiler logic to provide the check, meaning const stability is not blocked on const `TypeId` uncertainties.

Usage is also simpler, which means it serves a simplification purpose even after const `PartialEq`.

## Alternatives

- A lang feature that allows comparing types directly such as `T == i32`.
- The status quo `matches` from https://github.com/rust-lang/libs-team/issues/231.

## Links and related work

- C++ provides [`std::is_same`](https://en.cppreference.com/w/cpp/types/is_same).
- Zig has [`@typeInfo`](https://ziglang.org/documentation/master/#typeInfo) for reflection, the result of which can be compared.

## What happens now?

This issue contains an API change proposal (or ACP) and is part of the libs-api team [feature lifecycle]. Once this issue is filed, the libs-api team will review open proposals as capability becomes available. Current response times do not have a clear estimate, but may be up to several months.

[feature lifecycle]: https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html

## Possible responses

The libs team may respond in various different ways. First, the team will consider the *problem* (this doesn't require any concrete solution or alternatives to have been proposed):

- We think this problem seems worth solving, and the standard library might be the right place to solve it.
- We think that this probably doesn't belong in the standard library.

Second, if there's a concrete solution:

- We think this specific solution looks roughly right, approved, you or someone else should implement this. (Further review will still happen on the subsequent implementation PR.)
- We're not sure this is the right solution, and the alternatives or other materials don't give us enough information to be sure about that. Here are some questions we have that aren't answered, or rough ideas about alternatives we'd want to see discussed.

**Comment by tgross35** [comment]
This is really mostly a counterproposal to https://github.com/rust-lang/libs-team/issues/231, looking to address some of its limitations if this is to be a library feature. I do think that a lang feature allowing `T == U` should probably be considered because it seems easier to use, and could be extended to perform type narrowing without `const_downcast`.

Cc @KodrAus since most `TypeId` things are your design.

**Comment by RalfJung** [comment]
ISTM most of the time one of the types involved is dynamic, a la `&dyn Any`, and only the other type is statically known (like in the existing `downcast` methods)? How often are *both* types statically known?

**Comment by programmerjake** [comment]
I have encountered both types being statically known a few times e.g.
```rust
enum Select<T, F> {
    True(T),
    False(F),
}

trait Dispatch {
    type Type<const V: bool>: 'static;
}

pub fn select<T: Dispatch, const V: bool>(v: &mut T::Type<V>) -> Select<&mut T::Type<true>, &mut T::Type<false>> {
    if V {
        Select::True(<dyn Any>::downcast_mut(v).unwrap())
    } else {
        Select::False(<dyn Any>::downcast_mut(v).unwrap())
    }
}
```

**Comment by tgross35** [comment]
> ISTM most of the time one of the types involved is dynamic, a la `&dyn Any`, and only the other type is statically known (like in the existing `downcast` methods)? How often are _both_ types statically known?

I haven't seen any examples that seem to show this is the case in `const` functions. The example in https://github.com/rust-lang/libs-team/issues/231 and the UI test removed from https://github.com/rust-lang/rust/pull/99189 both use one generic and one specified type. Are there examples of this that couldn't be made to work by propagating a generic type, rather than passing a `TypdId`?

**Comment by NobodyXu** [comment]
There's a crate [castaway](https://docs.rs/castaway/) which has more generic API without having to specify the `'static` bound.

It uses a `LifetimeFree` trait to constraint the type to be cast, to avoid lifetime issues.

**Comment by RalfJung** [comment]
> Are there examples of this that couldn't be made to work by propagating a generic type, rather than passing a TypdId?

The typical example I was thinking of is code that handles `dyn Any` types, and then downcasts them back to their original type.

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting. We don't think this method will be very useful in practice because you often need to deal with a dynamic `TypeId`.

One alternative that we discussed was a `TypeId::is` method similar to the existing [`<dyn Any>::is`](https://doc.rust-lang.org/nightly/std/any/trait.Any.html#method.is) which checks if a `TypeId` refers to the given type.
